### PR TITLE
Adding rewrite rule set feature to AGIC via CRDs

### PR DESCRIPTION
**Pull request by paawankohli**
**Note: This is a draft PR**

## Checklist
- [x] The title of the PR is clear and informative
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] Issues addressed by the PR are mentioned in the description followed by `Fixes`.

## Description
This PR introduces a new CRD called agrewrite that can be used to configure application gateway rewrite rule sets within an AKS cluster that has AGIC installed. The rewrite rule set defined in the CRD can be referenced via the 'rewrite-rule-set-crd' annotation in the ingress manifest. 

## Fixes

This PR fixes https://github.com/Azure/application-gateway-kubernetes-ingress/issues/462, https://github.com/Azure/application-gateway-kubernetes-ingress/issues/1019, 

Related to https://github.com/Azure/application-gateway-kubernetes-ingress/issues/437, https://github.com/Azure/application-gateway-kubernetes-ingress/issues/1003

### CODE REVIEW

File: crds/AzureApplicationGatewayRewrite_v1beta1.yaml [hunk 1]
```diff
@@ -0,0 +1,109 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (akshaysngupta) [inline crds/AzureApplicationGatewayRewrite_v1beta1.yaml:1]: we don't need the v1beta1 crd

Comment (paawankohli) [reply crds/AzureApplicationGatewayRewrite_v1beta1.yaml]: bumped to v1 is subsequent commits

**Review by akshaysngupta — COMMENTED** [review]
Left comments

File: pkg/appgw/rewrites.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+	return appGwRewriteRuleSets
+}
+
+// getAGICRewriteRuleSets returns all the rewrite rule sets that are referenced in atleast one of the ingress manifests
+func (c appGwConfigBuilder) getAGICRewriteRuleSets(cbCtx *ConfigBuilderContext) []n.ApplicationGatewayRewriteRuleSet {
+
+	type rewriteCRDInfo struct {
+		ingressNamespace string
+		CRDName          string
```
Comment (akshaysngupta) [inline pkg/appgw/rewrites.go:56]: crdName

File: pkg/appgw/rewrites.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+	for _, rewriteRuleSetCRDInfo := range uniqueRewriteRuleSetCRD {
+
+		rewrite, err := c.k8sContext.GetRewrite(rewriteRuleSetCRDInfo.ingressNamespace, rewriteRuleSetCRDInfo.CRDName)
+
+		if err != nil {
+			klog.Errorf("Error occured while fetching rewrite CRD for %s. Please register CRD.", rewriteRuleSetCRDInfo.CRDName)
+			continue
+		}
+
+		appGwRewriteRuleSet = append(appGwRewriteRuleSet, c.makeRewrite(rewriteRuleSetCRDInfo.CRDName, rewrite))
```
Comment (akshaysngupta) [inline pkg/appgw/rewrites.go:90]: two rewrite rules CR in different namespace can have the same name. Can you use "namespace-name" as the name of the rewrite rule set ?

**Review by akshaysngupta — APPROVED** [review]
LGTM!

**Pull request merged**